ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/error/ErrorInterceptorSpec.scala

```diff
@@ -0,0 +1,325 @@
+  }
+
+  private def assertFooMissingError(
+      actualError: StatusRuntimeException,
+      expectedMsg: String,
+  ): Assertion = {
+    val actualStatus = actualError.getStatus
+    val actual = (actualStatus.getCode, actualStatus.getDescription)
+    val expectedDescription = s"FOO_MISSING_ERROR_CODE(11,0): Foo is missing: $expectedMsg"
+    val expected = (FooMissingErrorCode.category.grpcCode.get, expectedDescription)
+    actual shouldBe expected
+  }
+
+}
+
+object ErrorInterceptorSpec {
+
+  def server(tested: ErrorInterceptor, service: BindableService): ResourceOwner[Channel] = {
+    for {
+      server <- serverOwner(interceptor = tested, service = service)
+      channel <- GrpcClientResource.owner(Port(server.getPort))
+    } yield channel
+  }
+
+  private def serverOwner(
+      interceptor: ServerInterceptor,
+      service: BindableService,
+  ): ResourceOwner[Server] =
+    new ResourceOwner[Server] {
+      def acquire()(implicit context: ResourceContext): Resource[Server] =
+        Resource(Future {
+          val server =
+            NettyServerBuilder
+              .forAddress(new InetSocketAddress(InetAddress.getLoopbackAddress, 0))
+              .directExecutor()
+              .intercept(interceptor)
+              .addService(service)
+              .build()
+          server.start()
+          server
+        })(server => Future(server.shutdown().awaitTermination(10, TimeUnit.SECONDS): Unit))
+    }
+
+  object FooMissingErrorCode
+      extends ErrorCode(
+        id = "FOO_MISSING_ERROR_CODE",
+        ErrorCategory.InvalidGivenCurrentSystemStateResourceMissing,
+      )(ErrorClass.root()) {
+
+    case class Error(msg: String)(implicit
+        override val loggingContext: ContextualizedErrorLogger
+    ) extends BaseError.Impl(
+          cause = s"Foo is missing: $msg"
+        )
+
+  }
+
+  object FooUnknownErrorCode
+      extends ErrorCode(
+        id = "FOO_UNKNOWN_ERROR_CODE",
+        ErrorCategory.InvalidGivenCurrentSystemStateResourceMissing,
+      )(ErrorClass.root()) {
+
+    case class Error(msg: String)(implicit
+        override val loggingContext: ContextualizedErrorLogger
+    ) extends BaseError.Impl(
+          cause = s"Foo is unknown: $msg"
+        )
+
+  }
+
+  trait HelloService_Base extends BindableService {
+    self: HelloService =>
+
+    private val logger = ContextualizedLogger.get(getClass)
+    private val emptyLoggingContext = LoggingContext.newLoggingContext(identity)
+
+    implicit protected val damlLogger: DamlContextualizedErrorLogger =
+      new DamlContextualizedErrorLogger(logger, emptyLoggingContext, None)
+
+    override def bindService(): ServerServiceDefinition =
+      HelloServiceGrpc.bindService(this, scala.concurrent.ExecutionContext.Implicits.global)
+
+    override def fails(request: HelloRequest): Future[HelloResponse] = ??? // not used in this test
+  }
+
+  /** @param useSelfService       - whether to use self service error codes or a "rouge" exception
```

Comment:
```suggestion
  /** @param useSelfService       - whether to use self service error codes or a "rogue" exception
```

Reply:
Fix in the follow-up PR: https://github.com/digital-asset/daml/pull/12056